### PR TITLE
dns: enable dns failure refresh rate configuration

### LINKED ISSUE #7367: Add `dns_failure_refresh_rate` similar to `Cluster.dns_refresh_rate`
*Title*: *Should there be  `dns_failure_refresh_rate` similar to `Cluster.dns_refresh_rate` ?*

*Description*:
> So the `dns_refresh_rate` works exactly as expected. But when dns query returns no result, say our deployments are creating one and it takes some time but envoy got deployed before that. Now the DNS do not return any result. In this case envoy retries DNS only after the duration of `dns_refresh_rate`. Will it be meaningful for envoy to refresh DNS at a different rate when no results are returned ?

[optional *Relevant Links*:]
>https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cds.proto

### PR DESCRIPTION
**Pull request by venilnoronha**
Description: This adds a new `dns_failure_refresh_rate` configuration to the `Cluster` type to allow the configuration of the DNS refresh rate during failures (empty responses).
Risk Level: Low
Testing: Updated tests
Docs Changes: Updated docs
Release Notes: Added an entry describing the change
Fixes #7367

### CODE REVIEW

**Comment by junr03** [comment]
@snowp do you mind taking a look?

File: api/envoy/api/v2/cds.proto [hunk 1]
```diff
@@ -278,6 +278,16 @@ message Cluster {
...
 
+  // If the DNS failure refresh rate is specified and the cluster type is either
+  // :ref:`STRICT_DNS<envoy_api_enum_value_Cluster.DiscoveryType.STRICT_DNS>`,
+  // or :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`,
+  // this value is used as the cluster’s DNS refresh rate when requests are failing. If this setting
+  // is not specified, the value defaults to the DNS refresh rate. For cluster types other than
+  // :ref:`STRICT_DNS<envoy_api_enum_value_Cluster.DiscoveryType.STRICT_DNS>`
+  // and :ref:`LOGICAL_DNS<envoy_api_enum_value_Cluster.DiscoveryType.LOGICAL_DNS>`
+  // this setting is ignored.
+  google.protobuf.Duration dns_failure_refresh_rate = 42 [(validate.rules).duration.gt = {}];
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: This seems fine from an API review perspective. I'm a little late to this review though, I'm curious if we considered doing dynamic/adaptive back-off? There are other places we do this in Envoy today, e.g. see where `JitteredBackOffStrategy` is used.

Comment (mattklein123) [reply api/envoy/api/v2/cds.proto]: +1 that would probably be optimal, though would require a more complicated implementation including a cap on the back-off.

**Comment by venilnoronha** [comment]
The CI is failing for a very weird reason.

```
envoy/api/v2/cds.proto:703:34: Field number 42 has already been used in "envoy.api.v2.Cluster" by field "dns_failure_refresh_rate".
```

That's the new field I introduced in this PR, and it thinks the field existed already?

File: source/common/upstream/logical_dns_cluster.cc [hunk 2]
```diff
@@ -110,6 +113,8 @@ void LogicalDnsCluster::startResolve() {
             // checking, and creating real host connections.
             logical_host_->setNewAddress(new_address, lbEndpoint());
           }
+        } else {
+          refresh_rate = dns_failure_refresh_rate_ms_;
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/common/upstream/logical_dns_cluster.cc]: Addressed in 40ca4e2.

File: api/envoy/api/v2/cds.proto (hunk 1 above)
Comment (venilnoronha) [reply api/envoy/api/v2/cds.proto]: Addressed in 40ca4e2.

**Comment by mattklein123** [comment]
You have a merge conflict.

/wait

**Comment by mattklein123** [comment]
Actually I'm not sure what the issue is. Try merging master.

**Comment by venilnoronha** [comment]
@htuch I've addressed your comments in f7ab87b.

File: api/envoy/api/v2/cds.proto [hunk 3]
```diff
@@ -553,6 +553,35 @@ message Cluster {
...
+    google.protobuf.Duration base_interval = 1 [(validate.rules).duration = {
+      required: true,
+      gt: {nanos: 1000000}
+    }];
+
+    // Specifies the maximum interval between refreshes. This parameter is optional, but must be
+    // greater than or equal to the
+    // :ref:`base_interval <envoy_api_field_Cluster.RefreshRate.base_interval>`  if set. The default
+    // is 10 times the :ref:`base_interval <envoy_api_field_Cluster.RefreshRate.base_interval>`.
+    google.protobuf.Duration max_interval = 2 [(validate.rules).duration.gt = {seconds: 0}];
```
Comment (htuch) [inline api/envoy/api/v2/cds.proto]: Do you want a similar restriction here as with `base_interval`?

File: source/common/config/utility.h [hunk 4]
```diff
@@ -311,6 +312,16 @@ class Utility {
...
   }
+
+  /**
+   * Prepares the DNS failure refresh backoff strategy given the cluster configuration.
+   * @param cluster the cluster configuration.
+   * @param dns_refresh_rate_ms the default DNS refresh rate.
+   * @param random the random generator.
+   */
+  static BackOffStrategyPtr prepareDnsRefreshStrategy(const envoy::api::v2::Cluster& cluster,
+                                                      const uint64_t dns_refresh_rate_ms,
```
Comment (htuch) [inline source/common/config/utility.h]: Nit: `const` for a scalar doesn't do anything at the interface level.

File: source/common/config/utility.h [hunk 5]
```diff
@@ -311,6 +312,16 @@ class Utility {
                       name, filter_type));
     }
   }
+
+  /**
+   * Prepares the DNS failure refresh backoff strategy given the cluster configuration.
+   * @param cluster the cluster configuration.
+   * @param dns_refresh_rate_ms the default DNS refresh rate.
+   * @param random the random generator.
+   */
```
Comment (htuch) [inline source/common/config/utility.h]: Can you document `@return`?

**Comment by venilnoronha** [comment]
@htuch I've addressed your latest comments in adc0da2.

**Comment by venilnoronha** [comment]
unrelated flakes?

**Comment by lizan** [comment]
@venilnoronha No, you'll need to update stats_integration_test for the per cluster memory test when you add a new field to cluster.

**Pull request merged**